feat: rewrite Optimize API auth guide (#4315)
camunda/camunda-docs: optimize/apis-tools/optimize-api/optimize-api-authentication.md

description: "Connect business process-related event data and variable data held in external systems from third-party systems to Optimize, and more."
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

> All Optimize API requests except [the health readiness endpoint](./health-readiness.md) require authentication. To authenticate, generate a [JSON Web Token (JWT)](https://jwt.io/introduction/) and include it in each request.

Reviewer: ```suggestion
All Optimize API requests except the [health readiness](./health-readiness.md) endpoint require authentication. To authenticate, generate a [JSON Web Token (JWT)](https://jwt.io/introduction/) and include it in each request.
```
Suggestion to just have the endpoint name in the link.

> ## Generating a token

Reviewer: ```suggestion
## Generate a token
```
Suggestion to get away from using gerunds and make it more active, I think I suggested this in other reviews so not sure whether this was adopted - ignore if not, to keep consistent.

> ## Generating a token

Author: Ignoring for now for consistency; I'd like to get through all these rewrites (only one more after this!), and we can standardize this later.

<Tabs groupId="environment" defaultValue="saas" queryString values={
[
{label: 'SaaS', value: 'saas' },
{label: 'Self-Managed', value: 'self-managed' },
]}>
<TabItem value='saas'>

1. [Create client credentials]($docs$/guides/setup-client-connection-credentials/) in the **Clusters > Cluster name > API** tab of [Camunda Console](https://console.camunda.io/).
2. Add permissions to this client for **Optimize**.

> 3. Upon creating the client, capture the following values required to generate a token:

Author: https://github.com/camunda/camunda-docs/pull/4138#discussion_r1727666603

   <!-- this comment convinces the markdown processor to still treat the table as a table, but without adding surrounding paragraphs. 🤷 -->
   | Name                     | Environment variable name        | Default value                                |
   | ------------------------ | -------------------------------- | -------------------------------------------- |
   | Client ID                | `ZEEBE_CLIENT_ID`                | -                                            |
   | Client Secret            | `ZEEBE_CLIENT_SECRET`            | -                                            |
   | Authorization Server URL | `ZEEBE_AUTHORIZATION_SERVER_URL` | `https://login.cloud.camunda.io/oauth/token` |
   | Optimize REST Address    | `CAMUNDA_OPTIMIZE_BASE_URL`      | -                                            |
   <!-- this comment convinces the markdown processor to still treat the table as a table, but without adding surrounding paragraphs. 🤷 -->
   :::tip
   When client credentials are created, the `Client Secret` is only shown once. Save this `Client Secret` somewhere safe.
   :::

> 4. Execute an authentication request to the token issuer:

Author: No preference; let me know of any specific changes you'd like me to make in a follow-up.

   ```bash
   curl --request POST ${ZEEBE_AUTHORIZATION_SERVER_URL} \
       --header 'Content-Type: application/x-www-form-urlencoded' \
       --data-urlencode 'grant_type=client_credentials' \
       --data-urlencode 'audience=optimize.camunda.io' \
       --data-urlencode "client_id=${ZEEBE_CLIENT_ID}" \
       --data-urlencode "client_secret=${ZEEBE_CLIENT_SECRET}"
   ```

> 5. A successful authentication response looks like the following:

Author: This is a good call. Here's what it looks like currently: 

<img width="1012" alt="image" src="https://github.com/user-attachments/assets/26fd7ab6-a31c-4d74-a824-7d1f852d4c37">

That number 5 could definitely be a separate paragraph under number 4. I'll adjust these across all auth guides in a follow-up.

   ```json
   {
     "access_token": "<TOKEN>",
     "expires_in": 300,
     "refresh_expires_in": 0,
     "token_type": "Bearer",
     "not-before-policy": 0
   }
   ```
6. Capture the value of the `access_token` property and store it as your token.

</TabItem>

<TabItem value='self-managed'>

> 1. [Configure the `api.audience` setting](/self-managed/optimize-deployment/configuration/system-configuration.md#public-api) in your Optimize installation to match the audience property of the **Optimize API** [API in Identity]($docs$/self-managed/identity/user-guide/additional-features/adding-an-api/).

Author: This is different from other API Auth guides, and I don't feel great about requesting that someone re-configure their environment in order to get an auth token....but [as discussed in Slack](https://camunda.slack.com/archives/C55U06YRG/p1726686801515669), the out-of-the-box configuration of a Self-Managed instance doesn't actually work for client_credentials authentication.

> 1. [Configure the `api.audience` setting](/self-managed/optimize-deployment/configuration/system-configuration.md#public-api) in your Optimize installation to match the audience property of the **Optimize API** [API in Identity]($docs$/self-managed/identity/user-guide/additional-features/adding-an-api/).

Author: I don't know either, but if someone does and gives me specific changes, I'll take care of them in my follow-up.

3. [Add permissions to this application]($docs$/self-managed/identity/user-guide/additional-features/incorporate-applications/) for **Optimize API**.
4. Capture the `Client ID` and `Client Secret` from the application in Identity.
5. [Generate a token]($docs$/self-managed/identity/user-guide/authorizations/generating-m2m-tokens/) to access the REST API. Provide the `client_id` and `client_secret` from the values you previously captured in Identity.
   ```shell
   curl --location --request POST 'http://localhost:18080/auth/realms/camunda-platform/protocol/openid-connect/token' \
   --header 'Content-Type: application/x-www-form-urlencoded' \
   --data-urlencode "client_id=${CLIENT_ID}" \
   --data-urlencode "client_secret=${CLIENT_SECRET}" \
   --data-urlencode 'grant_type=client_credentials'
   ```
6. A successful authentication response looks like the following:
   ```json
   {
     "access_token": "<TOKEN>",
     "expires_in": 300,
     "refresh_expires_in": 0,
     "token_type": "Bearer",
     "not-before-policy": 0
   }
   ```
7. Capture the value of the `access_token` property and store it as your token.

:::note

> The Optimize API can also be configured in a Self-Managed environment to authenticate with a single shared access token. Refer to [Public API Configuration](/self-managed/optimize-deployment/configuration/system-configuration.md#public-api) for the configuration required to access the public API using a specific token.

Reviewer: ```suggestion
The Optimize API can also be configured in a Self-Managed environment to authenticate using a single shared access token. See [Public API Configuration](/self-managed/optimize-deployment/configuration/system-configuration.md#public-api) for the configuration required to access the public API using a specific token.
```

Author: I'll adjust this across all auth guides in a follow-up.

:::

</TabItem>

</Tabs>

## Using a token

Include the previously captured token as an authorization header in each request: `Authorization: Bearer <TOKEN>`.

> For example, to send a request to the Optimize API's ["Get dashboard IDs" endpoint](./dashboard/get-dashboard-ids.md):

Reviewer: ```suggestion
For example, to send a request to the Optimize API's ["Get dashboard IDs"](./dashboard/get-dashboard-ids.md) endpoint:
```
Also, should this endpoint name be in quotes?

Author: I don't know. I can make these changes in a follow-up, if we decide they're needed.
